[PLATFORM-622] Implement search panel minimise

timoxley opened the pull request:
![search-panel-minimise](https://user-images.githubusercontent.com/43438/57505486-1125f900-732b-11e9-8984-3b67892868ba.gif)

Behaviour works like so:

* If there's a search, it'll always show search results, regardless whether minimised or not.
* Minimised state can only be toggled by user clicking the minimise button.
* If there's no search & panel is minimised, it'll show nothing.
* If there's no search & panel is not minimised, it'll show some default content i.e. browse mode.

mondoreale commented on app/package.json:
```diff
@@ -169,7 +169,6 @@
     "react-ace": "^6.4.0",
     "react-avatar-editor": "^11.0.4",
     "react-calendar": "^2.17.5",
-    "react-children-addons": "^1.0.9",
```
👍

tumppi commented:
Noticed the same as Mariusz. We should probably disable resize when on search mode and there are no results.

mattatgit commented:
@juhah You mean when there are search results returned & displayed ? yeah, maybe thats a good idea.

timoxley commented:
* Fixed the glitch @mondoreale spotted above.
* Changed the way it handles locking resizing to the horizontal axis, now it only locks to horizontal when there's either no results, or the results fit in the current height. i.e. whenever the results are scrollable, you can change the height. This fixes an issue @juhah's raised prior about needing to clear results in order to resize the box.
* Also set it to restore the default height when switching from minimised to expanded, if the height is set to something less than the default.

timoxley commented:
@mondoreale please review

mondoreale commented:
@timoxley, I found something like this:

1. Have an empty canvas (freshly loaded).
2. Search for something the minimize.
3. Try to move the panel.
4. Each time you move it collapses/expands the panel (works like a toggle).

![May-16-2019 18-45-03](https://user-images.githubusercontent.com/320066/57871914-cadd0800-780a-11e9-8f10-78f23976f8dc.gif)

Can you reproduce?

mondoreale commented:
And something like this.

1. Search for something and minimize.
2. Hide the panel (`+` button in toolbar or `x` in module, both work)
3. Click `+` to show the panel again – it's expanded
4. Try resizing it – jumps back to minimized form

![May-16-2019 18-55-10](https://user-images.githubusercontent.com/320066/57872638-658a1680-780c-11e9-9082-f9daa0cbbd17.gif)

mondoreale commented:
And last one, the cursor.

1. Search for something and leave it, close the panel – cursor indicates horizontal resizing
2. Reopen the panel – cursor indicates resizing in both directions

![May-16-2019 19-05-57](https://user-images.githubusercontent.com/320066/57873244-f0b7dc00-780d-11e9-9e97-930ddd678d85.gif)

timoxley commented:
Ahh good finds @mondoreale  thanks

timoxley commented:
@mondoreale I've refactored to hooks and rewrote the resizing/minimising/autosizing behaviour. Please re-review, try break it.

mondoreale commented:
@timoxley, I'll try later today, or do it early in the morning tomorrow.

mondoreale commented on app/src/editor/shared/components/SearchPanel/index.jsx:
```diff
@@ -29,264 +28,266 @@ function MaybeDraggable({ disabled, children, ...props }) {
-                        posX: data.x,
-                        posY: data.y,
+    return (
+        <React.Fragment>
```
Is this `Fragment` required? `MaybeDraggable` is a single entry.

timoxley replied:
Fixed

mondoreale commented on app/src/editor/shared/components/SearchPanel/index.jsx:
```diff
@@ -29,264 +28,266 @@ function MaybeDraggable({ disabled, children, ...props }) {
+                                ref={scrollContainerRef}
+                            >
+                                <div ref={contentRef} role="listbox" className={styles.Content}>
+                                    {(() => {
```
How about we move it up and wrap with a `useMemo`?

mondoreale commented on app/src/editor/shared/components/SearchPanel/index.jsx:
```diff
@@ -29,264 +28,266 @@ function MaybeDraggable({ disabled, children, ...props }) {
-                        posY: data.y,
+    return (
+        <React.Fragment>
+            <MaybeDraggable
```
I wonder if we could use `Draggable`'s [`axis="none"`](https://github.com/mzabriskie/react-draggable#draggable-props) to disable d&d. Maybe we don't need another component.

timoxley replied:
Good thought, though currently the components that need this disabled keep it disabled for their entire lifetime, so saves the Draggable component from even instantiating if it's not going to be used.

timoxley commented on app/src/editor/shared/components/SearchPanel/index.jsx:
```diff
@@ -29,264 +28,266 @@ function MaybeDraggable({ disabled, children, ...props }) {
-        } = this.props
-        const { isExpanded, search, height } = this.state
-        const { current: currentContent } = this.contentRef
+    const toggleMinimize = () => (
```
Yeah sure, but no need, all of these are just passed directly to dom elements. Maybe worth doing for consistency I guess.

timoxley replied:
Added for consistency.

timoxley commented:
> It animates back in.

Yeah so this means we don't have to block resizing at all, you can always resize and it'll revert to autosize if the new size not currently applicable. IIRC it wasn't straightforward to accurately decide when it should block resizing and when it shouldn't.

> Another thing that is kinda strange is that the minimize toggle does nothing when search input isn't empty.

Yeah, but… what should it do?

timoxley commented:
Build passes, skipping deploy:

![image](https://user-images.githubusercontent.com/43438/58408596-b1587d80-80a0-11e9-9d66-9b2b0fddb614.png)
